### PR TITLE
[CPU] The Gather JIT kernel improvement. Part 1.

### PR DESCRIPTION
**Pull request by OleksiiZaderykhinCapgemini**
The next cases were implemented for the JIT kernel of the Gather operator.
The Blocked/Short case for static shapes for avx2 for 8-bit and 16-bit data types was implemented.
The Blocked/Short case for dynamic shapes for avx2 and avx512 for 8-bit, 16-bit, and 32-bit data types.

The implementation map for the JIT kernel for the Gather operation.
- avx512
    - dynamic shapes
        - Elementwise case
            - Short subcase  Implemented
            - Long subcase   Implemented
        - Blocked case
            - Short subcase  Implemented in this PR
            - Long subcase   Not implemented
    - static shapes
        - Elementwise case
            - Short subcase  Implemented
            - Long subcase   Implemented
        - Blocked case
            - Short subcase  Implemented
            - Long subcase   Not implemented
- avx2
    - dynamic shapes
        - Elementwise case
            - Short subcase  Implemented
            - Long subcase   Implemented
        - Blocked case
            - Short subcase  Implemented in this PR
            - Long subcase   Not implemented
    - static shapes
        - Elementwise case
            - Short subcase  Implemented
            - Long subcase   Implemented
        - Blocked case
            - Short subcase  Implemented in this PR for 8 bit and 16 bit data types, for 32 bit was implemented earlier
            - Long subcase   Not implemented
- SSE4.1                     Not implemented

This PR uses and depends on the StackAllocator from PR #13790 
This PR is a continuation of PR #13070  and is based on it.

### CODE REVIEW

File: src/plugins/intel_cpu/tests/functional/single_layer_tests/gather.cpp [hunk 1]
```diff
@@ -809,15 +1035,23 @@ INSTANTIATE_TEST_SUITE_P(smoke_static_4D_ref16, GatherLayerTestCPU,
                     ::testing::Values(additionalConfig[0])),
                 GatherLayerTestCPU::getTestCaseName);
 
+// TODO
+// This tests checks the Blocked/Short and the Blocked/Long cases if it use reference implementation
+// This checks fails currently because of the next.
+// Currently the Blocked/Short case was implemented in JIT kernel and is not use reference implementation anymore
+// But the Blocked/Long case has not been implemented in the JIT kernel yet.
+// And these tests must be uncommented after its implementation
+/*
```
Comment (nshchego) [inline src/plugins/intel_cpu/tests/functional/single_layer_tests/gather.cpp:1044]: Better to disable tests [here](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/tests/functional/shared_tests_instances/skip_tests_config.cpp)

File: src/plugins/intel_cpu/tests/functional/single_layer_tests/gather.cpp [hunk 2]
```diff
@@ -153,6 +153,221 @@ TEST_P(GatherLayerTestCPU, CompareWithRefs) {
...
+        Approach,
+        AfterAxisCase,
+        DataTypeBits,                       // Network precision
+        bool,                               // is dynamic shapes
+        int,                                // corner case index
+        CPUSpecificParams,                  // CPU specific params
+        std::map<std::string, std::string>  // Additional config
+> GatherCoverageLayerTestCPUParams;
+
+class GatherCoverageLayerTestCPU : public testing::WithParamInterface<GatherCoverageLayerTestCPUParams>,
```
Comment (nshchego) [inline src/plugins/intel_cpu/tests/functional/single_layer_tests/gather.cpp:182]: Can it be unified with GatherLayerTestCPU?

**Comment by nshchego** [comment]
There are already 2 PRs for Gather with a lot of refactoring, but the initial task not solved yet. Please focus on the main task. Each PR requires degradation/improvement validation. Some cases require manual validation, thus need to reduce a number of such PRs. If this refactoring is really needed to solve the issue move it into the first PR, but the second one should contain the final solution.

**Comment by OleksiiZaderykhinCapgemini** [comment]
> There are already 2 PRs for Gather with a lot of refactoring, but the initial task not solved yet. Please focus on the main task. Each PR requires degradation/improvement validation. Some cases require manual validation, thus need to reduce a number of such PRs. If this refactoring is really needed to solve the issue move it into the first PR, but the second one should contain the final solution.

@nshchego Ok then, I am closing this PR and will prepare PR with the final solution instead.

**Pull request closed without merging**